### PR TITLE
Dependency Update

### PR DESCRIPTION
**Pull request by rjurney**
This is my PR to update dependencies [cc @benedekrozemberczki and @GiulioRossetti @WhatTheFuzz @gkirgizov ]. You can see there is one error below that remains.

I get this error in DANMF that I need to figure out. I've tried twice before having to give up as I dug into scikit-leran's NMF class. Here goes again!

```python
============================================================== test session starts ==============================================================
platform linux -- Python 3.10.11, pytest-7.4.0, pluggy-1.2.0
rootdir: /home/rjurney/Software/karateclub
collected 6 items

test/community_detection_overlapping_test.py ..F...                                                                                       [100%]

=================================================================== FAILURES ====================================================================
__________________________________________________________________ test_danmf ___________________________________________________________________

    def test_danmf():
        """
        Test the DANMF procedure.
        """
        graph = nx.newman_watts_strogatz_graph(100, 5, 0.3)

        model = DANMF()

>       model.fit(graph)

test/community_detection_overlapping_test.py:128:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
karateclub/community_detection/overlapping/danmf.py:213: in fit
    self._pre_training()
karateclub/community_detection/overlapping/danmf.py:105: in _pre_training
    U, V = self._sklearn_pretrain(i)
karateclub/community_detection/overlapping/danmf.py:93: in _sklearn_pretrain
    U = nmf_model.fit_transform(self._Z)
../../anaconda3/envs/karateclub/lib/python3.10/site-packages/sklearn/utils/_set_output.py:140: in wrapped
    data_to_wrap = f(self, X, *args, **kwargs)
../../anaconda3/envs/karateclub/lib/python3.10/site-packages/sklearn/decomposition/_nmf.py:1568: in fit_transform
    self.reconstruction_err_ = _beta_divergence(
../../anaconda3/envs/karateclub/lib/python3.10/site-packages/sklearn/decomposition/_nmf.py:118: in _beta_divergence
    cross_prod = trace_dot((X * H.T), W)
../../anaconda3/envs/karateclub/lib/python3.10/site-packages/scipy/sparse/_base.py:584: in __mul__
    return self.multiply(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <100x100 sparse array of type '<class 'numpy.float64'>'
	with 636 stored elements in Compressed Sparse Row format>
other = array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.  ...  ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.14422184, 0.        ,
        0.        ]])

    def multiply(self, other):
        """Point-wise multiplication by another matrix, vector, or
        scalar.
        """
        # Scalar multiplication.
        if isscalarlike(other):
            return self._mul_scalar(other)
        # Sparse matrix or vector.
        if issparse(other):
            if self.shape == other.shape:
                other = self.__class__(other)
                return self._binopt(other, '_elmul_')
            # Single element.
            elif other.shape == (1, 1):
                return self._mul_scalar(other.toarray()[0, 0])
            elif self.shape == (1, 1):
                return other._mul_scalar(self.toarray()[0, 0])
            # A row times a column.
            elif self.shape[1] == 1 and other.shape[0] == 1:
                return self._mul_sparse_matrix(other.tocsc())
            elif self.shape[0] == 1 and other.shape[1] == 1:
                return other._mul_sparse_matrix(self.tocsc())
            # Row vector times matrix. other is a row.
            elif other.shape[0] == 1 and self.shape[1] == other.shape[1]:
                other = self._dia_container(
                    (other.toarray().ravel(), [0]),
                    shape=(other.shape[1], other.shape[1])
                )
                return self._mul_sparse_matrix(other)
            # self is a row.
            elif self.shape[0] == 1 and self.shape[1] == other.shape[1]:
                copy = self._dia_container(
                    (self.toarray().ravel(), [0]),
                    shape=(self.shape[1], self.shape[1])
                )
                return other._mul_sparse_matrix(copy)
            # Column vector times matrix. other is a column.
            elif other.shape[1] == 1 and self.shape[0] == other.shape[0]:
                other = self._dia_container(
                    (other.toarray().ravel(), [0]),
                    shape=(other.shape[0], other.shape[0])
                )
                return other._mul_sparse_matrix(self)
            # self is a column.
            elif self.shape[1] == 1 and self.shape[0] == other.shape[0]:
                copy = self._dia_container(
                    (self.toarray().ravel(), [0]),
                    shape=(self.shape[0], self.shape[0])
                )
                return copy._mul_sparse_matrix(other)
            else:
                raise ValueError("inconsistent shapes")

        # Assume other is a dense matrix/array, which produces a single-item
        # object array if other isn't convertible to ndarray.
        other = np.atleast_2d(other)

        if other.ndim != 2:
            return np.multiply(self.toarray(), other)
        # Single element / wrapped object.
        if other.size == 1:
            return self._mul_scalar(other.flat[0])
        # Fast case for trivial sparse matrix.
        elif self.shape == (1, 1):
            return np.multiply(self.toarray()[0, 0], other)

        ret = self.tocoo()
        # Matching shapes.
        if self.shape == other.shape:
            data = np.multiply(ret.data, other[ret.row, ret.col])
        # Sparse row vector times...
        elif self.shape[0] == 1:
            if other.shape[1] == 1:  # Dense column vector.
                data = np.multiply(ret.data, other)
            elif other.shape[1] == self.shape[1]:  # Dense matrix.
                data = np.multiply(ret.data, other[:, ret.col])
            else:
                raise ValueError("inconsistent shapes")
            row = np.repeat(np.arange(other.shape[0]), len(ret.row))
            col = np.tile(ret.col, other.shape[0])
            return self._coo_container(
                (data.view(np.ndarray).ravel(), (row, col)),
                shape=(other.shape[0], self.shape[1]),
                copy=False
            )
        # Sparse column vector times...
        elif self.shape[1] == 1:
            if other.shape[0] == 1:  # Dense row vector.
                data = np.multiply(ret.data[:, None], other)
            elif other.shape[0] == self.shape[0]:  # Dense matrix.
                data = np.multiply(ret.data[:, None], other[ret.row])
            else:
                raise ValueError("inconsistent shapes")
            row = np.repeat(ret.row, other.shape[1])
            col = np.tile(np.arange(other.shape[1]), len(ret.col))
            return self._coo_container(
                (data.view(np.ndarray).ravel(), (row, col)),
                shape=(self.shape[0], other.shape[1]),
                copy=False
            )
        # Sparse matrix times dense row vector.
        elif other.shape[0] == 1 and self.shape[1] == other.shape[1]:
            data = np.multiply(ret.data, other[:, ret.col].ravel())
        # Sparse matrix times dense column vector.
        elif other.shape[1] == 1 and self.shape[0] == other.shape[0]:
            data = np.multiply(ret.data, other[ret.row].ravel())
        else:
>           raise ValueError("inconsistent shapes")
E           ValueError: inconsistent shapes

../../anaconda3/envs/karateclub/lib/python3.10/site-packages/scipy/sparse/_compressed.py:471: ValueError
------------------------------------------------------------- Captured stdout call --------------------------------------------------------------
Pre-training layer 0.
Random state: 42
Number of iterations: 100
Z shape: (100, 100)
=============================================================== warnings summary ================================================================
test/community_detection_overlapping_test.py::test_egonet_splitter
test/community_detection_overlapping_test.py::test_nnsed
test/community_detection_overlapping_test.py::test_danmf
test/community_detection_overlapping_test.py::test_bigclam
test/community_detection_overlapping_test.py::test_mnmf
test/community_detection_overlapping_test.py::test_symmnmf
  /home/rjurney/Software/karateclub/karateclub/estimator.py:63: UserWarning: Please do be advised that the graph you have provided does not contain (some) edges in the main diagonal, for instance the self-loop constitued of (0, 0). These selfloops are necessary to ensure that the graph is traversable, and for this reason we create a copy of the graph and add therein the missing edges. Since we are creating a copy, this will immediately duplicate the memory requirements. To avoid this double allocation, you can provide the graph with the selfloops.
    warnings.warn(

test/community_detection_overlapping_test.py::test_danmf
  /home/rjurney/anaconda3/envs/karateclub/lib/python3.10/site-packages/sklearn/decomposition/_nmf.py:1665: ConvergenceWarning: Maximum number of iterations 100 reached. Increase it to improve convergence.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================ short test summary info ============================================================
FAILED test/community_detection_overlapping_test.py::test_danmf - ValueError: inconsistent shapes
==================================================== 1 failed, 5 passed, 7 warnings in 1.41s ====================================================
```

### CODE REVIEW

**Comment by rjurney** [comment]
Lol, ok. I haven't actually found any fixes yet other than some debug output :D Thought I was further along...

**Comment by rjurney** [comment]
Moar debugging:

```python
    def _sklearn_pretrain(self, i):
        """
        Pre-training a single layer of the model with sklearn.

        Arg types:
            * **i** *(int)* - The layer index.
        """

        print("Pre-training layer {}.".format(i))
        print(f"n_components: {self.layers[i]}")
        print("Random state: {}".format(self.seed))
        print("Number of iterations: {}".format(self.pre_iterations))
        print("Z shape: {}".format(self._Z.shape))

        nmf_model = NMF(
            n_components=self.layers[i],
            init="random",
            random_state=self.seed,
            max_iter=self.pre_iterations,
        )

        U = nmf_model.fit_transform(self._Z)
        V = nmf_model.components_
        return U, V
```

```
Pre-training layer 0.
n_components: 32
Random state: 42
Number of iterations: 100
Z shape: (100, 100)
```

**Comment by rjurney** [comment]
WE... HAVE... A... WINNER :) I downgraded to networkx `2.8.*` and I don't get an error but I do get this warning:

```python
  /home/rjurney/anaconda3/envs/karateclub/lib/python3.10/site-packages/networkx-2.8.8-py3.10.egg/networkx/linalg/graphmatrix.py:173: DeprecationWarning:

  The scipy.sparse array containers will be used instead of matrices
  in Networkx 3.0. Use `to_scipy_sparse_array` instead.
    return nx.to_scipy_sparse_matrix(G, nodelist=nodelist, dtype=dtype, weight=weight)

test/community_detection_overlapping_test.py::test_danmf
test/community_detection_overlapping_test.py::test_danmf
  /home/rjurney/Software/karateclub/karateclub/community_detection/overlapping/danmf.py:53: FutureWarning: adjacency_matrix will return a scipy.sparse array instead of a matrix in Networkx 3.0.
    self._A = nx.adjacency_matrix(

test/community_detection_overlapping_test.py::test_danmf
test/community_detection_overlapping_test.py::test_danmf
  <class 'networkx.utils.decorators.argmap'> compilation 16:4: FutureWarning: laplacian_matrix will return a scipy.sparse array instead of a matrix in Networkx 3.0.

test/community_detection_overlapping_test.py::test_danmf
test/community_detection_overlapping_test.py::test_danmf
  /home/rjurney/anaconda3/envs/karateclub/lib/python3.10/site-packages/sklearn/decomposition/_nmf.py:1665: ConvergenceWarning: Maximum number of iterations 100 reached. Increase it to improve convergence.
    warnings.warn(

test/community_detection_overlapping_test.py::test_mnmf
  /home/rjurney/Software/karateclub/karateclub/community_detection/overlapping/mnmf.py:82: FutureWarning: adjacency_matrix will return a scipy.sparse array instead of a matrix in Networkx 3.0.
    self._B1 = nx.adjacency_matrix(

test/community_detection_overlapping_test.py::test_symmnmf
  /home/rjurney/Software/karateclub/karateclub/community_detection/overlapping/symmnmf.py:61: FutureWarning: adjacency_matrix will return a scipy.sparse array instead of a matrix in Networkx 3.0.
    A = nx.adjacency_matrix(graph, nodelist=range(graph.number_of_nodes()))
```

We must cast it to an `np.ndarray` I think via `to_matrix()` :)

**Comment by rjurney** [comment]
Ok, I got this to work! All unit tests pass with networkx 3.3 :D

```
git commit -m "Made the calls in DANMF(Estimator)._setup_target_matrices to get the adjacency and Laplacian matrices convert the [now as of networkx 3.0+] sparse matrices returned into `np.matrices` by calling `.todense()`
```

**Comment by rjurney** [comment]
All unit tests pass. Closes #133 🥳 

Gentlemen [ @benedekrozemberczki and @GiulioRossetti @WhatTheFuzz @gkirgizov ] please check out this PR 
😁 🍇

**Comment by rjurney** [comment]
So I removed the remaining debug prints. The other minor changes are from `isort`, which aren't intentional. Should I remove them @benedekrozemberczki ?

File: karateclub/community_detection/overlapping/danmf.py [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
+from typing import Dict, List
+
 import networkx as nx
-from typing import List, Dict
+import numpy as np
 from sklearn.decomposition import NMF
+
 from karateclub.estimator import Estimator
 
+logger = logging.getLogger(__name__)
```
Comment (WhatTheFuzz) [inline karateclub/community_detection/overlapping/danmf.py:10]: Is this used anywhere?

File: karateclub/community_detection/overlapping/danmf.py [hunk 2]
```diff
@@ -1,9 +1,14 @@
-import numpy as np
```
Comment (WhatTheFuzz) [inline karateclub/community_detection/overlapping/danmf.py:1]: Any reason for moving the imports? pylint or similar?

Comment (rjurney) [reply karateclub/community_detection/overlapping/danmf.py:1]: @WhatTheFuzz `isort` made them PEP compliant. It goes system, PyPi then local, import them from, alphabetical. I forget which PEP standard. I could put them back.

**Comment by WhatTheFuzz** [comment]
Confirmed that tests pass with Python 3.12.0, pytest-7.4.2, and networkx 3.1.
```
47 passed, 96 warnings in 317.41s (0:05:17
```

File: karateclub/community_detection/overlapping/danmf.py (hunk 1 above)
Comment (rjurney) [reply karateclub/community_detection/overlapping/danmf.py]: @WhatTheFuzz nope, I just removed it.

File: karateclub/community_detection/overlapping/danmf.py (hunk 2 above)
Comment (rjurney) [reply karateclub/community_detection/overlapping/danmf.py:1]: @WhatTheFuzz I reverted them.

**Comment by rjurney** [comment]
@WhatTheFuzz I removed all the formatting stuff. I forget how to squash :)

File: karateclub/community_detection/overlapping/danmf.py (hunk 2 above)
Comment (WhatTheFuzz) [reply karateclub/community_detection/overlapping/danmf.py]: Ah sorry, I didn't mean to be prescriptive. I was just curious. I wasn't aware of the tool; I've typically just used `black`, which as I've just learned, uses that under the hood. Thanks for sharing.

Comment (rjurney) [reply karateclub/community_detection/overlapping/danmf.py]: No problem. @benedekrozemberczki I've used this branch of the library today with a class of 10 people and visualized the resulting samples with computed metrics and force directed layouts. The PR works :)

**Comment by rjurney** [comment]
@benedekrozemberczki can you plz merge?

**Comment by gkirgizov** [comment]
@benedekrozemberczki ping

**Comment by rjurney** [comment]
Hey @benedekrozemberczki can you give me admin privileges so I can merge pull requests?

**Comment by rjurney** [comment]
The library isn’t usable until I can get write permissions.

**Comment by rjurney** [comment]
@benedekrozemberczki ping

**Comment by LucaCappelletti94** [comment]
Will review PR and if everything checks out I will merge.

**Comment by LucaCappelletti94** [comment]
Confirm that test suite is passing with the updated version of NetworkX and edits check out. Merged.

**Pull request merged**